src/lavinmq/mqtt/retain_store.cr

```diff
@@ -0,0 +1,172 @@
+require "./topic_tree"
+require "digest/md5"
+
+module LavinMQ
+  module MQTT
+    class RetainStore
+      Log = LavinMQ::Log.for("retainstore")
+
+      MESSAGE_FILE_SUFFIX = ".msg"
+      INDEX_FILE_NAME     = "index"
+
+      alias IndexTree = TopicTree(String)
+
+      def initialize(@dir : String, @replicator : Clustering::Replicator, @index = IndexTree.new)
+        Dir.mkdir_p @dir
+        @files = Hash(String, File).new do |files, file_name|
+          file_path = File.join(@dir, file_name)
+          unless File.exists?(file_path)
+            File.open(file_path, "w").close
+          end
+          f = files[file_name] = File.new(file_path, "r+")
+          f.sync = true
+          f
+        end
+        @index_file = File.new(File.join(@dir, INDEX_FILE_NAME), "a+")
```

Comment:
We're doing `File.join(@dir, INDEX_FILE_NAME)` in multiple places. Should be enough to do once in the initializer and set an instance variable to the value that can be used in all other places.